File: felicia/examples/learn/service/ros_srv/cc/ros_srv_node_creator.cc
#include "felicia/core/felicia_init.h"
#include "felicia/core/master/master_proxy.h"
#include "felicia/examples/learn/service/common/cc/simple_service_flag.h"
#include "felicia/examples/learn/service/ros_srv/cc/ros_srv_client_node.h"
#include "felicia/examples/learn/service/ros_srv/cc/ros_srv_server_node.h"

namespace felicia {

int RealMain(int argc, char* argv[]) {
  FeliciaInit();

  SimpleServiceFlag delegate;
  FlagParser parser;
  parser.set_program_name("ros_srv_node_creator");
  if (!parser.Parse(argc, argv, &delegate)) {
    return 1;
  }

  MasterProxy& master_proxy = MasterProxy::GetInstance();
  Status s = master_proxy.Start();
  if (!s.ok()) {
    std::cerr << kRedError << s << std::endl;
    return 1;
  }

  NodeInfo node_info;
  node_info.set_name(delegate.name_flag()->value());

  if (delegate.is_server_flag()->value()) {
    master_proxy.RequestRegisterNode<RosSrvServerNode>(node_info, delegate);
  } else {
    master_proxy.RequestRegisterNode<RosSrvClientNode>(node_info, delegate);
  }

  master_proxy.Run();

  return 0;
}

}  // namespace felicia

int main(int argc, char* argv[]) { return felicia::RealMain(argc, argv); }

File: felicia/examples/learn/service/ros_srv/python/ros_srv_node_creator.py
import sys

import import_order_resolver  # Should be before import felicia_py
import felicia_py as fel
from felicia.core.protobuf.master_data_pb2 import NodeInfo
import felicia_py.command_line_interface as cli

from ros_srv_client_node import RosSrvClientNode
from ros_srv_server_node import RosSrvServerNode
from felicia.examples.learn.service.common.python.simple_service_flag import SimpleServiceFlag


def main():
    fel.felicia_init()

    delegate = SimpleServiceFlag()
    parser = cli.FlagParser()
    parser.set_program_name('ros_srv_node_creator')
    if not parser.parse(len(sys.argv), sys.argv, delegate):
        sys.exit(1)

    s = fel.MasterProxy.start()
    if not s.ok():
        print("{} {}.".format(cli.RED_ERROR, s), file=sys.stderr)
        sys.exit(1)

    node_info = NodeInfo()
    node_info.name = delegate.name_flag.value

    if delegate.is_server_flag.value:
        fel.MasterProxy.request_register_node(
            RosSrvServerNode, node_info, delegate)
    else:
        fel.MasterProxy.request_register_node(
            RosSrvClientNode, node_info, delegate)

    fel.MasterProxy.run()


if __name__ == '__main__':
    main()
